editor/components/block-list/block.js

```diff
@@ -522,7 +523,7 @@ export class BlockListBlock extends Component {
 					className="editor-block-list__block-edit"
 					data-block={ uid }
 				>
-
+					{ shouldShowContextualToolbar && <BlockContextualToolbar /> }
```

Comment:
This move was initially the key to unlocking collapsing margins. So I assumed that aspect would break when you moved this back. 

However later on I added https://github.com/WordPress/gutenberg/pull/7365/commits/3c0afac08b5cdbac268793329ffed27e9674d604#diff-00ebf096f7c483426f2a78d70d9cbd22R904, which as it turns out is sufficient to enable collapsing margins, even without being relocated.

However all the in-between styles were written to assume the position of the nested toolbar where it was before this move, so there were a ton of visual regressions as a result of this move, like toolbars not being in the right place, and so on. But I'm fixing those now, and in the end I think it was good to move this back. Thanks for all the help.